source/agora/common/BitMask.d

```diff
@@ -0,0 +1,195 @@
+
+public struct BitMask
+{
+    import std.algorithm;
+    import std.conv;
+    import std.range;
+
+    @safe pure:
+
+    /// String serialization
+    public string toString () const
+    {
+        return iota(this.count).map!(i => this[i] ? "1" : "0").join("");
+    }
+
+    /// Ditto
+    public static BitMask fromString (scope const(char)[] str)
+    {
+        auto bitmask = BitMask(str.length);
+        str.enumerate.each!((i, b)
+        {
+            if (b == '1')
+                bitmask[i] = true;
+        });
+        return bitmask;
+    }
+
+    nothrow:
+
+    /// Count of active validators who are expected / allowed to sign the block
+    public size_t count;
+
+    /// Bytes to hold bits to indicate if a validator has signed
+    public ubyte[] bytes;
+
+    public this (size_t count) inout
+    {
+        this.count = count;
+        if (count > 0)
+            this.bytes = new inout(ubyte)[1 + ((count - 1) / 8)];
+    }
+
+    public this (size_t count, in ubyte[] bytes)
+    {
+        this(count);
+        assert(this.bytes.length == bytes.length);
+        bytes.enumerate.each!((i, b) => this.bytes[i] = b);
+    }
+
+    // set the bits that are set in given BitMask
+    public auto addBitMask (in BitMask add)
+    {
+        assert(this.count == add.count, "BitMask assignment must be for same bit count");
+        iota(this.count).each!((size_t i)
+        {
+            if (add[i])
+                this[i] = true;
+        });
+        return this;
+    }
+
+    @nogc:
+
+    /// return the count of set bits
+    public size_t setCount () const
+    {
+        return iota(this.count).count!(i => this[i]);
+    }
+
+    // support setting a bit (asserts if trying to unset a bit)
+    public auto opIndexAssign(bool set, size_t bit_index)
+    {
+        if (bit_index >= this.count)
+                assert(0, "Attempt to set index beyond length of bitmask");
+        const size_t byte_index = (bit_index) / 8;
+        const size_t bit_index_in_byte = bit_index % 8;
+        if (set)
+            this.bytes[byte_index] |= mask(bit_index_in_byte);
+        else
+            assert(0, "Only setting bits is allowed!");
+        return this;
+    }
+
+    /// Gets a single bit's value
+    public bool opIndex (size_t bit_index) const
+    {
+        if (bit_index >= this.count)
+            assert(0, "Attempt to get index beyond length of bitmask");
+        return !!(this.bytes[bit_index / 8] & mask(bit_index));
+    }
+
+    /// Gets a bit mask which only includes a given index within a ubyte
+    pragma(inline, true)
+    private static ubyte mask (size_t index)
+    {
+        return (1 << (8 - 1 - (index % 8)));
+    }
+
+    /// Support for sorting by count of set bits not value
+    public int opCmp (in typeof(this) rhs) const
+    {
+        assert(this.count == rhs.count, "Comparing different sized BitMasks is not valid");
+        return this.setCount < rhs.setCount ? -1 : 1;
+    }
```

Comment:
Also need `opEquals`

Reply:
I think the default for this struct with length first and then `ubyte` array second is ok. Or maybe I am missing something?